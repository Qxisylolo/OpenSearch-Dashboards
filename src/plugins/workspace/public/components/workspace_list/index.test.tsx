/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { BehaviorSubject, of } from 'rxjs';
import { render, fireEvent, screen } from '@testing-library/react';
import { I18nProvider } from '@osd/i18n/react';
import { coreMock } from '../../../../../core/public/mocks';
import { navigateToWorkspaceDetail } from '../utils/workspace';
import { OpenSearchDashboardsContextProvider } from '../../../../../plugins/opensearch_dashboards_react/public';
import { WORKSPACE_USE_CASES } from '../../../common/constants';
import { WorkspaceList } from './index';

jest.mock('../utils/workspace');

jest.mock('../delete_workspace_modal', () => ({
  DeleteWorkspaceModal: ({ onClose }: { onClose: () => void }) => (
    <div aria-label="mock delete workspace modal">
      <button onClick={onClose} aria-label="mock delete workspace modal button" />
    </div>
  ),
}));

const formatDate = function (dateString: string) {
  const months = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec',
  ];

  const date = new Date(dateString);

  const month = months[date.getUTCMonth()];
  const day = String(date.getUTCDate()).padStart(2, '0');
  const year = date.getUTCFullYear();

  const hours = String(date.getUTCHours()).padStart(2, '0');
  const minutes = String(date.getUTCMinutes()).padStart(2, '0');
  const seconds = String(date.getUTCSeconds()).padStart(2, '0');
  const milliseconds = String(date.getUTCMilliseconds()).padStart(3, '0');

  return `${month} ${day},${year}@${hours}:${minutes}:${seconds}.${milliseconds}`;
};

function getWrapWorkspaceListInContext(
  workspaceList = [
    {
      id: 'id1',
      name: 'name1',
      features: ['use-case-all'],
      description:
        'should be able to see the description tooltip when hovering over the description',
      lastUpdatedTime: '1999-08-06T02:00:00.00Z',
    },
    {
      id: 'id2',
      name: 'name2',
      features: ['use-case-observability'],
      description:
        'should be able to see the description tooltip when hovering over the description',
      lastUpdatedTime: '1999-08-06T00:00:00.00Z',
    },
    {
      id: 'id3',
      name: 'name3',
      features: ['use-case-search'],
      description: '',
      lastUpdatedTime: '1999-08-06T01:00:00.00Z',
    },
  ],
  isDashboardAdmin = true
) {
  const coreStartMock = coreMock.createStart();
  coreStartMock.application.capabilities = {
    ...coreStartMock.application.capabilities,
    dashboards: {
      isDashboardAdmin,
    },
  };

  const services = {
    ...coreStartMock,
    workspaces: {
      workspaceList$: of(workspaceList),
    },
  };

  return (
    <I18nProvider>
      <OpenSearchDashboardsContextProvider services={services}>
        <WorkspaceList
          registeredUseCases$={new BehaviorSubject([WORKSPACE_USE_CASES.observability])}
        />
      </OpenSearchDashboardsContextProvider>
    </I18nProvider>
  );
}

describe('WorkspaceList', () => {
  it('should render title and table normally', () => {
    const { getByText, getByRole, container } = render(
      <WorkspaceList
        registeredUseCases$={new BehaviorSubject([WORKSPACE_USE_CASES.observability])}
      />
    );
    expect(getByText('Workspaces')).toBeInTheDocument();
    expect(getByRole('table')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  it('should render data in table based on workspace list data', async () => {
    const { getByText } = render(getWrapWorkspaceListInContext());

    // should display workspace names
    expect(getByText('name1')).toBeInTheDocument();
    expect(getByText('name2')).toBeInTheDocument();

    // should display use case
    expect(getByText('All use case')).toBeInTheDocument();
    expect(getByText('Observability')).toBeInTheDocument();
  });

  it('should be able to see the description tooltip when hovering over the description', async () => {
    const { getByText, getByTestId, queryByTestId } = render(getWrapWorkspaceListInContext());
    const description = getByText(
      'should be able to see the description tooltip when hovering over the description'
    );
    expect(description).toHaveClass('eui-textTruncate');
    fireEvent.mouseEnter(description);
    expect(getByTestId('workspaceList-hover-description')).toBeInTheDocument();

    const tooltip = await queryByTestId('workspaceList-hover-description');
    expect(tooltip).toBeInTheDocument();
    expect(tooltip).toHaveTextContent(
      'should be able to see the description tooltip when hovering over the description'
    );
    fireEvent.mouseLeave(description);
    expect(queryByTestId('workspaceList-hover-description')).not.toBeInTheDocument();
  });

  it('should be able to search and re-render the list', async () => {
    const { getByText, getByRole, queryByText } = render(getWrapWorkspaceListInContext());
    const input = getByRole('searchbox');
    fireEvent.change(input, {
      target: { value: 'name2' },
    });
    expect(getByText('name2')).toBeInTheDocument();
    expect(queryByText('name1')).not.toBeInTheDocument();
    expect(queryByText('name3')).not.toBeInTheDocument();
  });

  it('should be able to debounce when ', async () => {
    const { getByText, getByRole, queryByText } = render(getWrapWorkspaceListInContext());
    const input = getByRole('searchbox');
    fireEvent.change(input, {
      target: { value: 'name2' },
    });
    expect(getByText('name2')).toBeInTheDocument();
    expect(queryByText('name1')).not.toBeInTheDocument();
    expect(queryByText('name3')).not.toBeInTheDocument();
  });

  it('should be able to switch workspace after clicking name', async () => {
    const { getByText } = render(getWrapWorkspaceListInContext());
    const nameLink = getByText('name1');
    fireEvent.click(nameLink);
    expect(navigateToWorkspaceDetail).toBeCalled();
  });

  it('should be able to perform the time format transformation', async () => {
    const { getByText } = render(getWrapWorkspaceListInContext());
    expect(getByText(formatDate('1999-08-06T00:00:00.00Z'))).toBeInTheDocument();
    expect(getByText(formatDate('1999-08-06T01:00:00.00Z'))).toBeInTheDocument();
    expect(getByText(formatDate('1999-08-06T02:00:00.00Z'))).toBeInTheDocument();
  });

  // it('should be able to update workspace after clicking name', async () => {
  //   const { getAllByTestId } = render(getWrapWorkspaceListInContext());
  //   const editIcon = getAllByTestId('workspace-list-edit-icon')[0];
  //   fireEvent.click(editIcon);
  //   expect(navigateToWorkspaceDetail).toBeCalled();
  // });

  it('should be able to see the 3 operations: copy, update, delete after click in the meatballs button', async () => {
    const { getAllByTestId, getByText } = render(getWrapWorkspaceListInContext());
    const operationIcons = getAllByTestId('euiCollapsedItemActionsButton')[0];
    fireEvent.click(operationIcons);
    expect(getByText('Copy')).toBeInTheDocument();
    expect(getByText('Edit')).toBeInTheDocument();
    expect(getByText('Delete')).toBeInTheDocument();
  });

  it('should be able to update workspace after clicking name', async () => {
    const { getByText, getAllByTestId } = render(getWrapWorkspaceListInContext());
    const operationIcons = getAllByTestId('euiCollapsedItemActionsButton')[0];
    fireEvent.click(operationIcons);
    const editIcon = getByText('Edit');
    fireEvent.click(editIcon);
    expect(navigateToWorkspaceDetail).toBeCalled();
  });

  // it('should be able to copy workspace ID after clicking copy button', async () => {
  //   const { getByText, getAllByTestId } = render(getWrapWorkspaceListInContext());
  //   const operationIcons = getAllByTestId('euiCollapsedItemActionsButton')[0];
  //   fireEvent.click(operationIcons);
  //   const copyIcon = getByText('Copy');
  //   fireEvent.click(copyIcon);
  //   expect(
  //     navigator.clipboard.text
  //   ).toHaveBeenCalledWith('id1');
  // });

  it('should be able to call delete modal after clicking delete button', async () => {
    const { getByText, getAllByTestId } = render(getWrapWorkspaceListInContext());
    const operationIcons = getAllByTestId('euiCollapsedItemActionsButton')[0];
    fireEvent.click(operationIcons);
    const deleteIcon = getByText('Delete');
    fireEvent.click(deleteIcon);
    expect(screen.queryByLabelText('mock delete workspace modal')).toBeInTheDocument();
    const modalCancelButton = screen.getByLabelText('mock delete workspace modal button');
    fireEvent.click(modalCancelButton);
    expect(screen.queryByLabelText('mock delete workspace modal')).not.toBeInTheDocument();
  });

  it('should be able to pagination when clicking pagination button', async () => {
    const list = [
      {
        id: 'id1',
        name: 'name1',
        features: ['use-case-all'],
        description: '',
        lastUpdatedTime: '2024-08-06T00:00:00.00Z',
      },
      {
        id: 'id2',
        name: 'name2',
        features: ['use-case-observability'],
        description: '',
        lastUpdatedTime: '2024-08-06T00:00:00.00Z',
      },
      {
        id: 'id3',
        name: 'name3',
        features: ['use-case-search'],
        description: '',
        lastUpdatedTime: '2024-08-06T00:00:00.00Z',
      },
      {
        id: 'id4',
        name: 'name4',
        features: ['use-case-all'],
        description: '',
        lastUpdatedTime: '2024-08-06T00:00:00.00Z',
      },
      {
        id: 'id5',
        name: 'name5',
        features: ['use-case-observability'],
        description: '',
        lastUpdatedTime: '2024-08-06T00:00:00.00Z',
      },
      {
        id: 'id6',
        name: 'name6',
        features: ['use-case-search'],
        description: '',
        lastUpdatedTime: '2024-08-06T00:00:00.00Z',
      },
    ];
    const { getByTestId, getByText, queryByText } = render(getWrapWorkspaceListInContext(list));
    expect(getByText('name1')).toBeInTheDocument();
    expect(queryByText('name6')).not.toBeInTheDocument();
    const paginationButton = getByTestId('pagination-button-next');
    fireEvent.click(paginationButton);
    expect(queryByText('name1')).not.toBeInTheDocument();
    expect(getByText('name6')).toBeInTheDocument();
  });

  it('should display create workspace button for dashboard admin', async () => {
    const { getAllByText } = render(getWrapWorkspaceListInContext([], true));
    expect(getAllByText('Create workspace')[0]).toBeInTheDocument();
  });

  it('should hide create workspace button for non dashboard admin', async () => {
    const { queryByText } = render(getWrapWorkspaceListInContext([], false));
    expect(queryByText('Create workspace')).toBeNull();
  });

  // it('should be able to perform filtering when select a desired use case', async () => {
  //   const { getByText, getByRole, queryByText } = render(getWrapWorkspaceListInContext());

  //   // eslint-disable-next-line
  //   console.log(render(getWrapWorkspaceListInContext()));
  //   const filter = getByRole('selectfilters');
  //   fireEvent.change(filter, { target: { value: 'Observability' } });
  //   expect(getByText('name3')).toBeInTheDocument();
  //   expect(queryByText('name1')).not.toBeInTheDocument();
  //   expect(queryByText('name2')).not.toBeInTheDocument();
  // });
});
