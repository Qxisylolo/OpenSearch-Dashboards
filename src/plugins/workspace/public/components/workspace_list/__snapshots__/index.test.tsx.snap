// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkspaceList should render title and table normally 1`] = `
<div>
  <div
    class="euiPage euiPage--grow"
  >
    <div
      class="euiPanel euiPanel--paddingLarge euiPanel--borderRadiusNone euiPanel--plain euiPanel--hasShadow euiPageBody euiPageBody--borderRadiusNone euiPageBody--paddingLarge"
    >
      <header
        class="euiPageHeader euiPageHeader--responsive euiPage--restrictWidth-default euiPageHeader--center"
        style="padding-bottom: 0px; border-bottom: 0;"
      >
        <div
          class="euiPageHeaderContent"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--directionRow euiFlexGroup--responsive euiPageHeaderContent__top"
          >
            <div
              class="euiFlexItem"
            >
              <h1
                class="euiTitle euiTitle--large"
              >
                Workspaces
              </h1>
              <div
                class="euiSpacer euiSpacer--l"
              />
              <div
                class="euiText euiText--medium euiText--constrainedWidth"
              >
                <p>
                  Workspace allow you to save and organize library items, such as index patterns, visualizations, dashboards, saved searches, and share them with other OpenSearch Dashboards users. You can control which features are visible in each workspace, and which users and groups have read and write access to the library items in the workspace.
                </p>
              </div>
            </div>
          </div>
        </div>
      </header>
      <div
        class="euiPanel euiPanel--paddingLarge euiPanel--borderRadiusMedium euiPanel--plain euiPanel--noShadow euiPageContent euiPageContent--verticalCenter euiPageContent--horizontalCenter"
        role="main"
        style="width: 100%;"
      >
        <div>
          <div
            class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive euiFlexGroup--wrap"
          >
            <div
              class="euiFlexItem euiSearchBar__searchHolder"
            >
              <div
                class="euiFormControlLayout euiFormControlLayout--fullWidth"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <input
                    aria-label="This is a search bar. As you type, the results lower in the page will automatically filter."
                    class="euiFieldSearch euiFieldSearch--fullWidth"
                    placeholder="Search..."
                    type="search"
                    value=""
                  />
                  <div
                    class="euiFormControlLayoutIcons"
                  >
                    <span
                      class="euiFormControlLayoutCustomIcon"
                    >
                      <span
                        aria-hidden="true"
                        class="euiFormControlLayoutCustomIcon__icon"
                        data-euiicon-type="search"
                      />
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero euiSearchBar__filtersHolder"
            >
              <div
                class="euiFilterGroup"
              >
                <div
                  class="euiPopover euiPopover--anchorDownCenter"
                  id="field_value_selection_0"
                >
                  <div
                    class="euiPopover__anchor"
                  >
                    <button
                      class="euiButtonEmpty euiButtonEmpty--text euiFilterButton euiFilterButton--hasIcon"
                      type="button"
                    >
                      <span
                        class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                      >
                        <span
                          class="euiButtonContent__icon"
                          color="inherit"
                          data-euiicon-type="arrowDown"
                        />
                        <span
                          class="euiButtonEmpty__text"
                        >
                          <span
                            class="euiFilterButton__textShift"
                            data-text="Use Case"
                            title="Use Case"
                          >
                            Use Case
                          </span>
                        </span>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiSpacer euiSpacer--l"
          />
          <div
            class="euiBasicTable"
          >
            <div>
              <div
                class="euiTableHeaderMobile"
              >
                <div
                  class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
                >
                  <div
                    class="euiFlexItem euiFlexItem--flexGrowZero"
                  >
                    <div
                      class="euiCheckbox"
                    >
                      <input
                        aria-label="Select all rows"
                        class="euiCheckbox__input"
                        disabled=""
                        id="_selection_column-checkbox_generated-id"
                        type="checkbox"
                      />
                      <div
                        class="euiCheckbox__square"
                      />
                      <label
                        class="euiCheckbox__label"
                        for="_selection_column-checkbox_generated-id"
                      >
                        Select all rows
                      </label>
                    </div>
                  </div>
                  <div
                    class="euiFlexItem euiFlexItem--flexGrowZero"
                  >
                    <div
                      class="euiTableSortMobile"
                    >
                      <div
                        class="euiPopover euiPopover--anchorDownRight"
                      >
                        <div
                          class="euiPopover__anchor"
                        >
                          <button
                            class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--flushRight"
                            type="button"
                          >
                            <span
                              class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                            >
                              <span
                                class="euiButtonContent__icon"
                                color="inherit"
                                data-euiicon-type="arrowDown"
                              />
                              <span
                                class="euiButtonEmpty__text"
                              >
                                Sorting
                              </span>
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <table
                class="euiTable euiTable--responsive"
                id="__table_generated-id"
                tabindex="-1"
              >
                <caption
                  class="euiScreenReaderOnly euiTableCaption"
                />
                <thead>
                  <tr>
                    <th
                      class="euiTableHeaderCellCheckbox"
                      scope="col"
                    >
                      <div
                        class="euiTableCellContent"
                      >
                        <div
                          class="euiCheckbox euiCheckbox--inList euiCheckbox--noLabel"
                        >
                          <input
                            aria-label="Select all rows"
                            class="euiCheckbox__input"
                            data-test-subj="checkboxSelectAll"
                            disabled=""
                            id="_selection_column-checkbox_generated-id"
                            type="checkbox"
                          />
                          <div
                            class="euiCheckbox__square"
                          />
                        </div>
                      </div>
                    </th>
                    <th
                      aria-live="polite"
                      aria-sort="ascending"
                      class="euiTableHeaderCell"
                      data-test-subj="tableHeaderCell_name_0"
                      role="columnheader"
                      scope="col"
                      style="width: 25%;"
                    >
                      <button
                        class="euiTableHeaderButton euiTableHeaderButton-isSorted"
                        data-test-subj="tableHeaderSortButton"
                        type="button"
                      >
                        <span
                          class="euiTableCellContent"
                        >
                          <span
                            class="euiTableCellContent__text"
                            title="Name"
                          >
                            Name
                          </span>
                          <span
                            class="euiTableSortIcon"
                            data-euiicon-type="sortUp"
                          />
                        </span>
                      </button>
                    </th>
                    <th
                      class="euiTableHeaderCell"
                      data-test-subj="tableHeaderCell_useCase_1"
                      role="columnheader"
                      scope="col"
                      style="width: 20%;"
                    >
                      <span
                        class="euiTableCellContent"
                      >
                        <span
                          class="euiTableCellContent__text"
                          title="Use case"
                        >
                          Use case
                        </span>
                      </span>
                    </th>
                    <th
                      class="euiTableHeaderCell"
                      data-test-subj="tableHeaderCell_description_2"
                      role="columnheader"
                      scope="col"
                      style="width: 20%;"
                    >
                      <span
                        class="euiTableCellContent"
                      >
                        <span
                          class="euiTableCellContent__text"
                          title="Description"
                        >
                          Description
                        </span>
                      </span>
                    </th>
                    <th
                      class="euiTableHeaderCell"
                      data-test-subj="tableHeaderCell_lastUpdatedTime_3"
                      role="columnheader"
                      scope="col"
                      style="width: 25%;"
                    >
                      <span
                        class="euiTableCellContent"
                      >
                        <span
                          class="euiTableCellContent__text"
                          title="Last updated"
                        >
                          Last updated
                        </span>
                      </span>
                    </th>
                    <th
                      class="euiTableHeaderCell"
                      role="columnheader"
                      scope="col"
                    >
                      <span
                        class="euiTableCellContent euiTableCellContent--alignRight"
                      >
                        <span
                          class="euiTableCellContent__text"
                          title="Actions"
                        >
                          Actions
                        </span>
                      </span>
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr
                    class="euiTableRow"
                  >
                    <td
                      class="euiTableRowCell euiTableRowCell--isMobileFullWidth"
                      colspan="6"
                    >
                      <div
                        class="euiTableCellContent euiTableCellContent--alignCenter"
                      >
                        <span
                          class="euiTableCellContent__text"
                        >
                          <div
                            class="euiEmptyPrompt"
                          >
                            <span
                              color="subdued"
                              data-euiicon-type="spacesApp"
                            />
                            <div
                              class="euiSpacer euiSpacer--m"
                            />
                            <h3
                              class="euiTitle euiTitle--small"
                            >
                              No workspace available
                            </h3>
                            <span
                              class="euiTextColor euiTextColor--subdued"
                            >
                              <div
                                class="euiSpacer euiSpacer--m"
                              />
                              <div
                                class="euiText euiText--medium"
                              >
                                There are no workspace to display. Create workspace to get started.
                              </div>
                            </span>
                          </div>
                        </span>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
